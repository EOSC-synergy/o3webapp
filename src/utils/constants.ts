/**
 * A module defining all constants that are shared between different components.
 * It contains default values and settings for those components.
 *
 * @module Constants */ // used for auto generation of JSDocs with better-docs

/**
 * Stores the default model group.
 *
 * Modify this object to change the name or the models that appear when
 * the WebApp is initially loaded.
 * @category Utils
 */
export const DEFAULT_MODEL_GROUP = {
    groupId: undefined, // no valid id => add new group
    groupName: 'Example Group',
    modelList: [
        'CCMI-1_ACCESS_ACCESS-CCM-refC2',
        'CCMI-1_ACCESS_ACCESS-CCM-senC2fGHG',
        'CCMI-1_CCCma_CMAM-refC2',
    ],
};

/** Stores the error message that is shown if no month is selected.
 * @category Utils
 */
export const NO_MONTH_SELECTED =
    'Invalid Selection: Selecting at least one month should make this error go away.';

/** Stores all the months in short form. (first three letters)
 * @category Utils
 */
export const months = [
    Symbol('Jan'),
    Symbol('Feb'),
    Symbol('Mar'),
    Symbol('Apr'),
    Symbol('May'),
    Symbol('Jun'),
    Symbol('Jul'),
    Symbol('Aug'),
    Symbol('Sep'),
    Symbol('Oct'),
    Symbol('Nov'),
    Symbol('Dec'),
];

/** Stores the amount of months in a season
 * @category Utils
 */
export const NUM_MONTHS_IN_SEASON = 3;

/** Stores the amount of months in a year
 * @category Utils
 */
export const NUM_MONTHS = 12;

export type Latitude = {
    minLat: number;
    maxLat: number;
};

export type LatitudeBand = {
    text: symbol;
    value: Latitude;
};

/** Stores the latitude bands and its min/max values
 * @category Utils
 */
export const latitudeBands: LatitudeBand[] = [
    {
        text: Symbol('Southern Hemisphere (SH) Polar (90–60°S)'),
        value: { minLat: -90, maxLat: -60 },
    },
    {
        text: Symbol('SH Mid-Latitudes (60–35°S)'),
        value: { minLat: -60, maxLat: -35 },
    },
    {
        text: Symbol('Tropics (20°S–20°N)'),
        value: { minLat: -20, maxLat: 20 },
    },
    {
        text: Symbol('Northern Hemisphere (NH) Mid-Latitudes (35–60°N)'),
        value: { minLat: 35, maxLat: 60 },
    },
    {
        text: Symbol('NH Polar (60–90°N)'),
        value: { minLat: 60, maxLat: 90 },
    },
    {
        text: Symbol('Near-Global (60°S–60°N)'),
        value: { minLat: -60, maxLat: 60 },
    },
    {
        text: Symbol('Global (90°S–90°N)'),
        value: { minLat: -90, maxLat: 90 },
    },
    // !!! Custom must be last in array !!!
    {
        text: Symbol('Custom'),
        // ts-expect-error TODO: figure out this
        value: 'custom' as unknown as Latitude,
    },
];

/**
 * An "enum" which stores the plot types as they come from the api.
 * @constant {object}
 */
export const O3AS_PLOTS = {
    // used for internal testing or manual if-else
    tco3_zm: 'tco3_zm',
    tco3_return: 'tco3_return',
} as const;
export type O3AS_PLOTS = typeof O3AS_PLOTS[keyof typeof O3AS_PLOTS];

/**
 * The start year, no data is fetched for years before this point in time
 * @category Utils
 */
export const START_YEAR = 1960;

/**
 * The end year, no data is fetched for years after this point in time
 * @category Utils
 */
export const END_YEAR = 2100;

// important for api data transformation
/** This String stores the description of the antarctic region
 * @category Utils
 */
const SH_POLAR = 'SH Polar';

/** This string stores the description of the south hemisphere region
 * @category Utils
 */
const SH_MID = 'SH mid-lat';

/** This string stores the description of the tropics region
 * @category Utils
 */
const TROPICS = 'Tropics';

/** This string stores the description of the north hemisphere region
 * @category Utils
 */
const NH_MID = 'NH mid-lat';

/** This string stores the description of the arctic region
 * @category Utils
 */
const NH_POLAR = 'NH Polar';

/** This string stores the description of the near global region
 * @category Utils
 */
const NEAR_GLOBAL = 'Near global';

/** This string stores the description of the global region
 * @category Utils
 */
const GLOBAL = 'Global';

/** This string stores the description of the user region
 * @category Utils
 */
export const USER_REGION = 'User region';

// user region must be last element
export const ALL_REGIONS_ORDERED = [
    SH_POLAR,
    SH_MID,
    TROPICS,
    NH_MID,
    NH_POLAR,
    NEAR_GLOBAL,
    GLOBAL,
    USER_REGION,
];

/**
 * The mean: this appears in the model group card and is used to identify its statistical value settings.
 * @category Utils
 */
export const mean = 'mean';
/**
 * The standard deviation: this appears in the model group card and is used to identify its statistical value settings.
 * @category Utils
 */
export const std = 'standard deviation';
/**
 * The median: this appears in the model group card and is used to identify its statistical value settings.
 * @category Utils
 */
export const median = 'median';
/**
 * The percentile: this appears in the model group card and is used to identify its statistical value settings.
 * @category Utils
 */
export const percentile = 'percentile';

/**
 * The statistical values that are computable are listed here as
 * an "enum"
 * @category Utils
 */
export const STATISTICAL_VALUES = {
    [mean]: mean,
    [std]: std,
    [median]: median,
    [percentile]: percentile,
} as const;
export type STATISTICAL_VALUES = typeof STATISTICAL_VALUES[keyof typeof STATISTICAL_VALUES];

/**
 * Name for the helper statistical value series: This line is used to calculate the mean+/-std
 * @category Utils
 */
export const stdMean = 'stdMean';

/**
 * The lower percentile ~15.87th
 * @category Utils
 */
export const lowerPercentile = 'lowerPercentile';

/**
 * The lower percentile ~84.13th
 * @category Utils
 */
export const upperPercentile = 'upperPercentile';

/**
 * This extended sv list is used in the calculation to handle some changes that
 * came up after the specification / design phase: standard deviation is not one
 * but TWO lines. Same goes for percentile which is lower- AND upper-percentile.
 * @category Utils
 */
export const EXTENDED_SV_LIST = [stdMean, lowerPercentile, upperPercentile];

/**
 * The same statistical values as a list to verify certain payload data.
 * @category Utils
 */
export const STATISTICAL_VALUES_LIST = Object.values(STATISTICAL_VALUES);

/**
 * The default color of the backgrounds (Navbar, Footer, Sidebar, ...)
 * @category Utils
 */
export const BACKGROUND_BASE_COLOR = '#111';
