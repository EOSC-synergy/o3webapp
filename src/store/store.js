import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
