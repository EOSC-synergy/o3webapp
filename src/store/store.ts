import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
