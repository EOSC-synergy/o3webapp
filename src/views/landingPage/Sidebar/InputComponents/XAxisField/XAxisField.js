import React from "react";
import {useDispatch, useSelector} from "react-redux";
import {setDisplayXRange, selectPlotXRange} from "../../../../../store/plotSlice/plotSlice";
import {Typography, Grid, TextField, FormControl} from "@mui/material";
import {START_YEAR, END_YEAR} from '../../../../../utils/constants';

/**
 * Enables the user to choose the range that should be visible on the x-axis of the plot.
 * @component
 * @param {Object} props
 * @param {function} props.reportError function for error handling
 * @returns {JSX.Element} a jsx containing two text-fields and labels
 */
function XAxisField(props) {
    /**
     * A dispatch function to dispatch actions to the redux store.
     */
    const dispatch = useDispatch();

    /**
     * An object containing the minX and maxX values for the x-axis.
     */
    const {years: {minX, maxX}} = useSelector(selectPlotXRange);

    /**
     * Stores the minX and maxX values
     * and checks their validation before sending it to the Redux store.
     */
    const [stateX, setStateX] = React.useState({minX, maxX});

    /**
     * Handles the change of the minimum value.
     *
     * @param event the input value
     */
    const handleChangeMin = (event) => {
        if (!isNaN(event.target.value)) {
            setStateX({minX: event.target.value, maxX: maxX});
            if (event.target.value >= START_YEAR && event.target.value <= END_YEAR && event.target.value <= maxX) {
                dispatch(setDisplayXRange({years: {minX: parseInt(event.target.value), maxX: maxX}}));
            }
        }
    }

    /**
     * Handles the change of the maximum value.
     *
     * @param event the input value
     */
    const handleChangeMax = (event) => {
        if (!isNaN(event.target.value)) {
            setStateX({minX: minX, maxX: event.target.value});
            if (event.target.value >= START_YEAR && event.target.value <= END_YEAR && minX <= event.target.value) {
                dispatch(setDisplayXRange({years: {minX: minX, maxX: parseInt(event.target.value)}}));
            }
        }
    }

    return (
        <Grid container sx={{width: "90%", marginLeft: "auto", marginRight: "auto", marginTop: "5%"}}>
            <Grid item xs={3}>
                <Typography>X-Axis:</Typography>
            </Grid>
            <Grid item xs={3} sx={{mt: "-8px"}}>
                <FormControl sx={{width: '85%'}}>
                    <TextField
                        variant="outlined"
                        id="outlined-basic"
                        size="small"
                        value={stateX.minX}
                        onChange={handleChangeMin}
                        error={stateX.minX < START_YEAR || stateX.minX > END_YEAR || stateX.minX >= maxX}
                        helperText={stateX.minX < START_YEAR ? `<${START_YEAR}` : (stateX.minX > END_YEAR ? `>${END_YEAR}` : (stateX.minX >= maxX ? `min>=max` : ''))}
                    />
                </FormControl>
            </Grid>
            <Grid item xs={1} sx={{mt: "-5px"}}>
                <h2 style={{display: "inline"}}> - </h2>
            </Grid>
            <Grid item xs={3} sx={{mt: "-8px"}}>
                <FormControl sx={{width: '85%'}}>
                    <TextField
                        variant="outlined"
                        id="outlined-basic"
                        size="small"
                        value={stateX.maxX}
                        onChange={handleChangeMax}
                        error={stateX.maxX < START_YEAR || stateX.maxX > END_YEAR || minX >= stateX.maxX}
                        helperText={stateX.maxX < START_YEAR ? `<${START_YEAR}` : (stateX.maxX > END_YEAR ? `>${END_YEAR}` : (minX >= stateX.maxX ? `min>=max` : ''))}
                    />
                </FormControl>
            </Grid>
        </Grid>
    );
}

export default XAxisField;