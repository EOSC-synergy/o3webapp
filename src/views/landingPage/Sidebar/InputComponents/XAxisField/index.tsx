import React, { ChangeEvent, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    setDisplayXRange,
    selectPlotXRange,
    YearsBasedXRange,
} from '../../../../../store/plotSlice/plotSlice';
import { Typography, Grid, TextField, FormControl } from '@mui/material';
import { START_YEAR, END_YEAR } from '../../../../../utils/constants';

/**
 * Enables the user to choose the range that should be visible on the x-axis of the plot.
 * @component
 * @returns {JSX.Element} a jsx containing two text-fields and labels
 */
const XAxisField: React.FC = () => {
    /**
     * A dispatch function to dispatch actions to the redux store.
     * @constant {function}
     */
    const dispatch = useDispatch();

    /**
     * An object containing the minX and maxX values for the x-axis.
     * @see {@link selectPlotXRange}
     * @constant {Object}
     */
    const {
        years: { minX, maxX },
    } = useSelector(selectPlotXRange) as YearsBasedXRange;

    /**
     * Stores the minX and maxX values
     * and checks their validation before sending it to the Redux store.
     * @constant {Array}
     */
    const [stateX, setStateX] = React.useState({ minX, maxX });

    /**
     * Handles the change of the minimum value.
     *
     * @see {@link setStateX}
     * @see {@link setDisplayXRange}
     * @param event {Event} event that triggered the call of this function
     */
    const handleChangeMin = (event: ChangeEvent<HTMLInputElement>) => {
        if (!isNaN(event.target.valueAsNumber)) {
            setStateX({ minX: event.target.valueAsNumber, maxX: maxX });
            if (
                event.target.valueAsNumber >= START_YEAR &&
                event.target.valueAsNumber <= END_YEAR &&
                event.target.valueAsNumber < maxX
            ) {
                dispatch(
                    setDisplayXRange({ years: { minX: event.target.valueAsNumber, maxX: maxX } })
                );
            }
        }
    };

    useEffect(() => {
        setStateX({ minX, maxX });
    }, [minX, maxX]);

    /**
     * Handles the change of the maximum value.
     *
     * @param event {Event} event that triggered the call of this function
     * @see {@link setStateX}
     * @see {@link setDisplayXRange}
     */
    const handleChangeMax = (event: ChangeEvent<HTMLInputElement>) => {
        if (!isNaN(event.target.valueAsNumber)) {
            setStateX({ minX: minX, maxX: event.target.valueAsNumber });
            if (
                event.target.valueAsNumber >= START_YEAR &&
                event.target.valueAsNumber <= END_YEAR &&
                minX < event.target.valueAsNumber
            ) {
                dispatch(
                    setDisplayXRange({ years: { minX: minX, maxX: event.target.valueAsNumber } })
                );
            }
        }
    };

    return (
        <Grid container sx={{ width: '90%', marginLeft: 'auto', marginRight: 'auto' }}>
            <Grid item xs={3}>
                <Typography>X-Axis:</Typography>
            </Grid>
            <Grid item xs={3} sx={{ mt: '-8px' }}>
                <FormControl sx={{ width: '85%' }}>
                    <TextField
                        variant="outlined"
                        id="outlined-basic"
                        size="small"
                        value={stateX.minX}
                        onChange={handleChangeMin}
                        error={
                            stateX.minX < START_YEAR ||
                            stateX.minX > END_YEAR ||
                            stateX.minX >= maxX
                        }
                        helperText={
                            stateX.minX < START_YEAR
                                ? `<${START_YEAR}`
                                : stateX.minX > END_YEAR
                                ? `>${END_YEAR}`
                                : stateX.minX >= maxX
                                ? `min>=max`
                                : ''
                        }
                        inputProps={{ 'data-testid': 'XAxisField-left-input' }}
                    />
                </FormControl>
            </Grid>
            <Grid item xs={1} sx={{ mt: '-5px' }}>
                <h2 style={{ display: 'inline' }}> - </h2>
            </Grid>
            <Grid item xs={3} sx={{ mt: '-8px' }}>
                <FormControl sx={{ width: '85%' }}>
                    <TextField
                        variant="outlined"
                        id="outlined-basic"
                        size="small"
                        value={stateX.maxX}
                        onChange={handleChangeMax}
                        error={
                            stateX.maxX < START_YEAR ||
                            stateX.maxX > END_YEAR ||
                            minX >= stateX.maxX
                        }
                        helperText={
                            stateX.maxX < START_YEAR
                                ? `<${START_YEAR}`
                                : stateX.maxX > END_YEAR
                                ? `>${END_YEAR}`
                                : minX >= stateX.maxX
                                ? `min>=max`
                                : ''
                        }
                        inputProps={{ 'data-testid': 'XAxisField-right-input' }}
                    />
                </FormControl>
            </Grid>
        </Grid>
    );
};

export default XAxisField;
