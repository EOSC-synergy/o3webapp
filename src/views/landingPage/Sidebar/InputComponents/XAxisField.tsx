import React, { ChangeEventHandler, FC, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { selectPlotId, selectPlotXRangeZm, setDisplayXRangeForPlot } from 'store/plotSlice';
import { Grid, TextField } from '@mui/material';
import { END_YEAR, START_YEAR } from 'utils/constants';
import invariant from 'tiny-invariant';

const XAxisField: FC = () => {
    const dispatch = useDispatch();

    const activePlot = useSelector(selectPlotId);
    invariant(activePlot === 'tco3_zm', 'XAxisField is only available for tco3_zm plot');

    const {
        years: { minX, maxX },
    } = useSelector(selectPlotXRangeZm);

    const [xRange, setXRange] = useState({ minX, maxX });

    const handleChangeMin: ChangeEventHandler<HTMLInputElement> = (event) => {
        const value = event.target.valueAsNumber;

        if (!isNaN(value)) {
            setXRange({ minX: value, maxX: maxX });
            if (value >= START_YEAR && value <= END_YEAR && value < maxX) {
                dispatch(
                    setDisplayXRangeForPlot({
                        plotId: activePlot,
                        displayXRange: {
                            years: {
                                minX: value,
                                maxX: maxX,
                            },
                        },
                    })
                );
            }
        }
    };

    useEffect(() => {
        setXRange({ minX, maxX });
    }, [minX, maxX]);

    const handleChangeMax: ChangeEventHandler<HTMLInputElement> = (event) => {
        const value = event.target.valueAsNumber;
        if (!isNaN(value)) {
            setXRange({ minX: minX, maxX: value });
            if (value >= START_YEAR && value <= END_YEAR && minX < value) {
                dispatch(
                    setDisplayXRangeForPlot({
                        plotId: activePlot,
                        displayXRange: {
                            years: {
                                minX: minX,
                                maxX: value,
                            },
                        },
                    })
                );
            }
        }
    };

    return (
        <Grid container sx={{ justifyContent: 'center' }}>
            <Grid xs={5}>
                <TextField
                    variant="outlined"
                    size="small"
                    value={xRange.minX}
                    onChange={handleChangeMin}
                    error={
                        xRange.minX < START_YEAR || xRange.minX > END_YEAR || xRange.minX >= maxX
                    }
                    helperText={
                        xRange.minX < START_YEAR
                            ? `must be >${START_YEAR}`
                            : xRange.minX > END_YEAR
                            ? `must be <${END_YEAR}`
                            : xRange.minX >= maxX
                            ? `must be lower than maximum`
                            : ''
                    }
                    inputProps={{ 'data-testid': 'XAxisField-left-input' }}
                    type="number"
                    label="X Minimum"
                />
            </Grid>
            <Grid
                xs={1}
                sx={{
                    marginTop: '.5rem',
                    display: 'flex',
                    justifyContent: 'center',
                }}
            >
                -
            </Grid>
            <Grid xs={5}>
                <TextField
                    variant="outlined"
                    size="small"
                    value={xRange.maxX}
                    onChange={handleChangeMax}
                    error={
                        xRange.maxX < START_YEAR || xRange.maxX > END_YEAR || minX >= xRange.maxX
                    }
                    helperText={
                        xRange.maxX < START_YEAR
                            ? `must be >${START_YEAR}`
                            : xRange.maxX > END_YEAR
                            ? `must be <${END_YEAR}`
                            : minX >= xRange.maxX
                            ? `must be higher than minimum`
                            : ''
                    }
                    inputProps={{ 'data-testid': 'XAxisField-right-input' }}
                    type="number"
                    label="X Maximum"
                />
            </Grid>
        </Grid>
    );
};

export default XAxisField;
