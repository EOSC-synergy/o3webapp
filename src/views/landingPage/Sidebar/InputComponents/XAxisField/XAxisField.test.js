import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux'
import XAxisField from './XAxisField';
import { createTestStore } from "../../../../../store/store"
import {fireEvent} from '@testing-library/dom';
import { END_YEAR, START_YEAR } from '../../../../../utils/constants';

describe('tests basic rendering', () => {
    let store;
    beforeEach(() => {
        store = createTestStore();
    });

    it('renders without crashing', () => {
        render(<>
            <Provider store={store}>
                <XAxisField reportError={() => {}} />
            </Provider>
        </>)
    });

    it('renders correctly', () => {

        const { container } = render(<>
            <Provider store={store}>
                <XAxisField reportError={() => {}} />
            </Provider>
        </>);

        expect(container).toMatchSnapshot();
    });
});

describe('test functionality redux', () => {
    let store;
    beforeEach(() => {
        store = createTestStore();
    });

    it('should update the store on user input (left field)', () => {
        // value before change
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.minX).toEqual(1960);

        const { getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputField = getByTestId("XAxisField-left-input");
        fireEvent.change(inputField, {target: {value: "2000"}}); // change input
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.minX).toEqual(2000);
    });

    it('should update the store on user input (right field)', () => {
        // value before change
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX).toEqual(2100);

        const { getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputField = getByTestId("XAxisField-right-input");
        fireEvent.change(inputField, {target: {value: "2010"}}); // change input
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX).toEqual(2010);
    });
});

describe('test error handling functionality', () => {
    let store;
    beforeEach(() => {
        store = createTestStore();
    });

    it("doesn't update if the value is smaller than START_YEAR (1960)", () => {
        const previousState = store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.minX;

        const { container, getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputField = getByTestId("XAxisField-left-input");
        fireEvent.change(inputField, {target: {value: START_YEAR - 1}}); // change input
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.minX).toEqual(previousState); // no update took place
        expect(container).toHaveTextContent(`<${START_YEAR}`); // expect that error is displayed
    });

    it("doesn't update if the value is greater than END_YEAR (1960)", () => {
        const previousState = store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX;

        const { container, getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputField = getByTestId("XAxisField-right-input");
        fireEvent.change(inputField, {target: {value: END_YEAR + 1}}); // change input
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX).toEqual(previousState); // no update took place
        expect(container).toHaveTextContent(`>${END_YEAR}`); // expect that error is displayed
    });

    it("doesn't update if LEFT >= RIGHT", () => {
        const previousState = store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX;

        const { container, getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputFieldLeft = getByTestId("XAxisField-left-input");
        const inputFieldRight = getByTestId("XAxisField-right-input");
        fireEvent.change(inputFieldLeft, {target: {value: 2060}}); // change input
        fireEvent.change(inputFieldRight, {target: {value: 2060}}); // change input (should have no effect)
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.minX).toEqual(2060); // assure left got updated correctly
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX).toEqual(previousState); // no update took place on the right side
        expect(container).toHaveTextContent("min>=max"); // expect that error is displayed
    });

    it("doesn't accept non-numerical input", () => {
        const previousState = store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX;

        const { getByTestId } = render(
            <Provider store={store}>
                <XAxisField />
            </Provider>
        );     
        const inputFieldRight = getByTestId("XAxisField-right-input");
        fireEvent.change(inputFieldRight, {target: {value: "test"}}); // change input (should have no effect)
        expect(store.getState().plot.plotSpecificSettings.tco3_zm.displayXRange.years.maxX).toEqual(previousState); // no update took place
    });

});