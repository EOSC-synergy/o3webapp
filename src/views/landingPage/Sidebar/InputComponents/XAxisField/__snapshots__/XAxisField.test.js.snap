// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests basic rendering renders correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container css-wts08w-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        X-Axis:
      </p>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1lrjc51-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root css-olnp1t-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="XAxisField-left-input"
              id="outlined-basic"
              type="text"
              value="1960"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-nnbavb"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1rbsaog-MuiGrid-root"
    >
      <h2
        style="display: inline;"
      >
         - 
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1lrjc51-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root css-olnp1t-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="XAxisField-right-input"
              id="outlined-basic"
              type="text"
              value="2100"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-nnbavb"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
